Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs

```diff
@@ -81,27 +82,32 @@ public TcpSocketClient(EncryptionManager encryptionManager, bool keepAlive, bool
             }
         }
 
-        private async Task Connect(Uri uri)
+        private async Task Connect(Uri uri, TimeSpan timeOut)
         {
-            var addresses = await uri.ResolveAsync(_ipv6Enabled);
-            AggregateException innerErrors = null;
-            for (var i = 0; i < addresses.Length; i++)
+            using (CancellationTokenSource cancellationSource = new CancellationTokenSource(timeOut))
             {
-                try
-                {
-                    await _client.ConnectAsync(addresses[i], uri.Port).ConfigureAwait(false);
-                    return;
-                }
-                catch (Exception e)
+                var addresses = await uri.ResolveAsync(_ipv6Enabled);
+                AggregateException innerErrors = null;
+                for (var i = 0; i < addresses.Length; i++)
                 {
-                    var error = new IOException($"Failed to connect to server '{uri}' via IP address '{addresses[i]}': {e.Message}", e);
-                    innerErrors = innerErrors == null ? new AggregateException(error) : new AggregateException(innerErrors, error);
+                    try
+                    {
+                        cancellationSource.Token.ThrowIfCancellationRequested();
```

Comment:
Before making any changes to the current behaviour, we should clarify what we will do when a single socket connection times out in the for-loop. Do we cancel all the connection process (throw an exception) or will we continue trying the next IPAddress available? 

It depends on how "establishing a connection" is conceived on the developer side. Is it the maximum time allowed to get a working SocketConnection or is it the maximum time to connect to an IPAddress as a pure Socket - which on failure an actual working SocketConnection could take some multitude of the timeout value given.

Keeping it as it is for not introducing a behavioral change. Will work on it when we clarify its behavior-to-be.